Refacto geojson parsing and converting features

gchoqueux opened the pull request:
## Description
**Refacto `GeojsonParser,` `Feature2Mesh` and `feature2texture`** 

- add merge features option (so the convert creates only one mesh) 
- Use one array of number to store the coordinates

**BREAKING CHANGE**
One array of number to store the coordinates instead of array of coordinates

## Motivation and Context
- **fix #731 bad performance**

   - Before bug from PR #691, total time by functions are :
     - ` geojsonParser.parse`  320 ms
     - `feature2Mesh`  216 ms
         `parse + convert` 536 ms
     - `renderScene` 350 ms
     - **total** 866 ms

   - From introduction of bug with PR #691, total time by functions are :
     - ` geojsonParser.parse`  874 ms (x 2.7 slower)
     - `feature2Mesh`  451 ms (x 2.1 slower)
        `parse + convert ` 1325 ms (x 2.5 slower)
     - `renderScene` 885 ms (x 2.5 slower)
     - **total** 2210 ms (x 2.5 slower)

   - After fix from this PR, total time by functions are :
     - ` geojsonParser.parse`  90 ms (x 10 faster)
     - `feature2Mesh`  115 ms (x 4 faster)
        `parse + convert ` 205 ms (x 6.5 faster)
     - `renderScene` 314 ms (x 2.8 faster)
     - **total** 724 ms (x 3 faster)

- **ref: Improving the Feature type #763**

gchoqueux committed:
fix(core): Cache return mesh with null geometry

The Object Removal Helper causes inconsistencies
when it assigns null to a geometry present in the Cache.
Actually, the cache can provide a mesh whose geometry is null.
see https://github.com/iTowns/itowns/issues/869 #869

gliegard commented:
Great performance fix !!

gchoqueux commented:
i must fix GUI pixel and line size

zarov commented on src/Parser/GeoJsonParser.js:
```diff
@@ -159,58 +169,121 @@ function readGeometry(feature, crsIn, crsOut, geometry, filteringExtent, options
+                f.vertices.crs = crsOut;
+                res.features.push(f);
+                if (options.buildExtent) {
+                    mergeExtent(res, f.extent);
```
Not necessary, it should already be at the maximum extent at this point, no ?

gchoqueux replied:
With `options.mergeFeatures`, there are finally 3 features (one multiline, multipoint and multipolygon) with one extent for each feature. But, It's necessary to compute the global extent, which is the merger of the 3 extents.

zarov replied:
Ok, missed a condition above that !

zarov commented on src/Parser/GeoJsonParser.js:
```diff
@@ -29,118 +21,136 @@ function readCRS(json) {
 // Multi-* geometry types are merged in one.
+const coords = new Coordinates('EPSG:4978', 0, 0, 0);
+// filter with the first point
+const firstPtIsOut = (extent, aCoords, crs) => {
```
Filtering on the first point is kind of arbitrary, what can you propose instead ?

gchoqueux replied:
It's a choice. But we must filter the features that are partially on the extent. We added, later, adding filtering options.

zarov replied:
Ok if it's overwritten, but we can improve this point in the future for sure.

zarov commented on src/Provider/DataSourceProvider.js:
```diff
@@ -47,14 +47,17 @@ function fetchData(url, format, networkOptions, extentSource) {
         origin: layer.source.origin,
+        mergeFeatures: layer.mergeFeatures === undefined ? true : layer.mergeFeatures,
+        withNormal: layer.type === 'geometry',
+        withAltitude: layer.type === 'geometry',
```
What if I have a plane view with no altitude but still geometry data to show ?

gchoqueux replied:
you still need an altitude of zero

zarov replied:
Yes

zarov commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -41,23 +44,25 @@ function fillColorArray(colors, length, r, g, b, offset) {
-        // convert coordinate to position
-        coordinate.xyz(vec);
+const coord = new Coordinates('EPSG:4326', 0, 0);
+function coordinatesToVertices(ptsIn, normals, target, altitude = 0, extrude = 0, offsetOut = 0, countIn = ptsIn.length / 3, startIn = offsetOut) {
```
Please quickly complete the documentation above

gchoqueux replied:
added

zarov commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -41,23 +44,25 @@ function fillColorArray(colors, length, r, g, b, offset) {
+    const endIn = startIn + countIn;
+    for (let i = startIn, j = offsetOut; i < endIn; i += 3, j += 3) {
+        let a = 0;
+        a = !isNaN(altitude) ? altitude : altitude({}, coord.set(ptsIn.crs, ptsIn[i], ptsIn[i + 1], ptsIn[i + 2]));
```
No more support for an array of altitude ? We may need to find a solution to have both a value, an array or a function for those properties (in another issue).

gchoqueux replied:
support added

zarov commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -79,168 +84,223 @@ function coordinatesToVertices(contour, altitude, target, offset = 0, extrude =
-            for (let j = start; j < end; j++) {
-                indices.push(j);
-                indices.push(j + 1);
+            if (start > 0xffff) {
```
Why can't it go further ?

gchoqueux replied:
To avoid integer overflow with indice value (16 bits), I added comment

zarov replied:
Ok, maybe you can add a warning that the geometry is too large or something like that

zarov commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -79,168 +84,223 @@ function coordinatesToVertices(contour, altitude, target, offset = 0, extrude =
-    const altitude = getProperty('altitude', options, 0, properties, feature.vertices);
-    const color = getProperty('color', options, randomColor, properties);
+var lineMaterial = new THREE.LineBasicMaterial({ vertexColors: THREE.VertexColors });
+function featureToLine(feature, options) {
```
Sidenote: we should move to `THREE.MeshLine` to have correct line width support under Angle.

gchoqueux replied:
yes, I agree

zarov commented on src/Renderer/ThreeExtended/Feature2Mesh.js:
```diff
@@ -79,168 +84,223 @@ function coordinatesToVertices(contour, altitude, target, offset = 0, extrude =
 }
 
+function area(contour, offset, count) {
+    // Fix me: not the same result with THREE.shapeUtils.area
```
This needs to be fixed before merging

gchoqueux replied:
sorry, it was old comment, it was already fixed

zarov commented on src/Renderer/ThreeExtended/Feature2Texture.js:
```diff
@@ -71,17 +74,21 @@ function drawPoint(ctx, vertice, origin, scale, style = {}) {
     for (const geometry of feature.geometry) {
+        const properties = geometry.properties;
+
+        let gStyle = style;
```
This could be reverted outside of the loop

gchoqueux replied:
that is to say? 
`properties` may change each step

zarov replied:
Yes indeed if `style` is a function, but you can declare `gStyle = style` outside the loop

gchoqueux replied:
yes

zarov commented on test/unit/geojson.js:
```diff
@@ -16,12 +16,12 @@ function parse(geojson) {
 describe('GeoJsonParser', function () {
```
Could you add some tests with maybe more features, but testing with all the new options or not in it ?

gchoqueux replied:
tests unit added

zarov reviewed (changes requested):
Good work, it is going quite faster, and it is appreciated !

gchoqueux commented:
Branch is rebased

zarov reviewed (approved):
Seems good to me, thanks for this big fix

gchoqueux committed:
refacto(parser/converter): merge features option and use one array of number to store the coordinates

* fix bad performance #731
* ref: Improving the Feature type #763

BREAKING CHANGE

One array of number to store the coordinates instead of array of coordinates
